File: test/runtime_fixture_ffi.mjs
import { Error as GleamError } from "./gleam.mjs";
import { existsSync, unlinkSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { spawn, spawnSync } from "node:child_process";

const descendantMarker = join(
  tmpdir(),
  `kangaroo-isolate-descendant-${globalThis.process.pid}.marker`,
);
const parallelBarrierPrefix = join(
  tmpdir(),
  `kangaroo-parallel-batch-${globalThis.process.pid}`,
);

function parallelMarker(side) {
  return `${parallelBarrierPrefix}-${side}.marker`;
}

export function reset_parallel_barrier() {
  for (const side of ["left", "right"]) {
    const marker = parallelMarker(side);
    if (existsSync(marker)) unlinkSync(marker);
  }
}

export async function parallel_barrier(side) {
  const other = side === "left" ? "right" : "left";
  writeFileSync(parallelMarker(side), "ready");
  const deadline = Date.now() + 5000;
  while (!existsSync(parallelMarker(other))) {
    if (Date.now() >= deadline) {
      throw new Error("scheduled JavaScript modules did not overlap");
    }
    await new Promise((resolve) => setTimeout(resolve, 5));
  }
}

export function promise_pass() {
  return Promise.resolve(undefined);
}

export function promise_reject() {
  return Promise.reject(new Error("async rejected"));
}

export function promise_never() {
  return new Promise(() => {});
}

export function left_value() {
  return 1;
}

export function right_value() {
  return 2;
}

export function error_result() {
  return new GleamError("not an integer");
}

export function left_string() {
  return "same\nold";
}

export function right_string() {
  return "same\nnew";
}

export function non_binary_assert() {
  const error = new Error("synthetic non-binary assert");
  error.gleam_error = "assert";
  error.kind = "guard";
  error.left = { value: 1 };
  error.right = { value: 2 };
  throw error;
}

export function spawn_descendant() {
  const code = `setTimeout(() => require("node:fs").writeFileSync(${JSON.stringify(descendantMarker)}, "survived"), 100);`;
  const parentCode = `const { spawn } = require("node:child_process"); spawn(process.execPath, ["-e", ${JSON.stringify(code)}], { stdio: "ignore" }); setTimeout(() => {}, 1000);`;
  const denoCode = `await new Promise(resolve => setTimeout(resolve, 100)); await Deno.writeTextFile(${JSON.stringify(descendantMarker)}, "survived");`;
  const denoParentCode = `new Deno.Command(Deno.execPath(), { args: ["eval", ${JSON.stringify(denoCode)}], stdout: "null", stderr: "null" }).spawn(); await new Promise(resolve => setTimeout(resolve, 1000));`;
  const deno = typeof globalThis.Deno !== "undefined";
  spawn(
    globalThis.process.execPath,
    deno ? ["eval", denoParentCode] : ["-e", parentCode],
    { stdio: "ignore" },
  );
}

export function spawn_cleanup_race() {
  spawn_descendant();
}

export function spawn_orphan_descendant() {
  const deno = typeof globalThis.Deno !== "undefined";
  const writer = deno
    ? `await new Promise(resolve => setTimeout(resolve, 400)); await Deno.writeTextFile(${JSON.stringify(descendantMarker)}, "survived");`
    : `setTimeout(() => require("node:fs").writeFileSync(${JSON.stringify(descendantMarker)}, "survived"), 400);`;
  const parent = deno
    ? `new Deno.Command(Deno.execPath(), { args: ["eval", ${JSON.stringify(writer)}], stdout: "null", stderr: "null" }).spawn().unref();`
    : `require("node:child_process").spawn(process.execPath, ["-e", ${JSON.stringify(writer)}], { stdio: "ignore" }).unref();`;
  return new Promise((resolve, reject) => {
    const child = spawn(
      globalThis.process.execPath,
      deno ? ["eval", parent] : ["-e", parent],
      { stdio: "ignore" },
    );
    child.once("error", reject);
    child.once("close", resolve);
  });
}

export function spawn_native_orphan_descendant() {
  if (typeof globalThis.Deno !== "undefined") {
    const writer = `await new Promise(resolve => setTimeout(resolve, 100)); await Deno.writeTextFile(${JSON.stringify(descendantMarker)}, "survived");`;
    const parent = `new Deno.Command(Deno.execPath(), { args: ["eval", ${JSON.stringify(writer)}], stdout: "null", stderr: "null" }).spawn().unref();`;
    return new globalThis.Deno.Command(globalThis.Deno.execPath(), {
      args: ["eval", parent],
      stdout: "null",
      stderr: "null",
    }).output();
  }
  if (typeof globalThis.Bun !== "undefined") {
    const writer = `setTimeout(() => require("node:fs").writeFileSync(${JSON.stringify(descendantMarker)}, "survived"), 100);`;
    const parent = `Bun.spawn([process.execPath, "-e", ${JSON.stringify(writer)}], { stdout: "ignore", stderr: "ignore" }).unref();`;
    return globalThis.Bun.spawn(
      [globalThis.process.execPath, "-e", parent],
      { stdout: "ignore", stderr: "ignore" },
    ).exited;
  }
  return Promise.resolve(undefined);
}

export function spawn_port_orphan_descendant() {
  return undefined;
}

export function kill_output_collector() {}

export function kill_test_owner_from_link() {}

export function exit_test_worker() {
  globalThis.process.exit(7);
}

export function spawn_native_descendant() {
  const code = `setTimeout(() => require("node:fs").writeFileSync(${JSON.stringify(descendantMarker)}, "survived"), 100);`;
  if (typeof globalThis.Bun !== "undefined") {
    globalThis.Bun.spawn([globalThis.process.execPath, "-e", code], {
      stdout: "ignore",
      stderr: "ignore",
    });
    return;
  }
  if (typeof globalThis.Deno !== "undefined") {
    const denoCode = `await new Promise(resolve => setTimeout(resolve, 100)); await Deno.writeTextFile(${JSON.stringify(descendantMarker)}, "survived");`;
    new globalThis.Deno.Command(globalThis.Deno.execPath(), {
      args: ["eval", denoCode],
      stdout: "null",
      stderr: "null",
    }).spawn();
    return;
  }
  spawn(globalThis.process.execPath, ["-e", code], { stdio: "ignore" });
}

export function complete_native_child() {
  const code = `require("node:fs").writeFileSync(${JSON.stringify(descendantMarker)}, "completed");`;
  if (typeof globalThis.Bun !== "undefined") {
    return globalThis.Bun.spawn([globalThis.process.execPath, "-e", code], {
      stdin: "ignore",
      stdout: "ignore",
      stderr: "ignore",
    }).exited;
  }
  if (typeof globalThis.Deno !== "undefined") {
    const denoCode = `await Deno.writeTextFile(${JSON.stringify(descendantMarker)}, "completed");`;
    return new globalThis.Deno.Command(globalThis.Deno.execPath(), {
      args: ["eval", denoCode],
      stdin: "null",
      stdout: "null",
      stderr: "null",
    }).output();
  }
  return new Promise((resolve, reject) => {
    const child = spawn(globalThis.process.execPath, ["-e", code], {
      stdio: "ignore",
    });
    child.once("error", reject);
    child.once("exit", code => code === 0
      ? resolve()
      : reject(new Error(`native child exited ${code}`)));
  });
}

export function spawn_synchronous_descendant() {
  if (typeof globalThis.Deno !== "undefined") {
    const writer = `await new Promise(resolve => setTimeout(resolve, 100)); await Deno.writeTextFile(${JSON.stringify(descendantMarker)}, "survived");`;
    const parent = `new Deno.Command(Deno.execPath(), { args: ["eval", ${JSON.stringify(writer)}], stdout: "null", stderr: "null" }).spawn();`;
    new globalThis.Deno.Command(globalThis.Deno.execPath(), {
      args: ["eval", parent],
      stdout: "null",
      stderr: "null",
    }).outputSync();
    return;
  }
  const writer = `setTimeout(() => require("node:fs").writeFileSync(${JSON.stringify(descendantMarker)}, "survived"), 100);`;
  const parent = `require("node:child_process").spawn(process.execPath, ["-e", ${JSON.stringify(writer)}], { stdio: "ignore" }).unref();`;
  if (typeof globalThis.Bun !== "undefined") {
    globalThis.Bun.spawnSync([globalThis.process.execPath, "-e", parent], {
      stdout: "ignore",
      stderr: "ignore",
    });
    return;
  }
  spawnSync(globalThis.process.execPath, ["-e", parent], { stdio: "ignore" });
}

export function native_output_timeout() {
  const nodeCode = `setTimeout(() => require("node:fs").writeFileSync(${JSON.stringify(descendantMarker)}, "survived"), 200);`;
  if (typeof globalThis.Deno !== "undefined") {
    const denoCode = `await new Promise(resolve => setTimeout(resolve, 200)); await Deno.writeTextFile(${JSON.stringify(descendantMarker)}, "survived");`;
    return new globalThis.Deno.Command(globalThis.Deno.execPath(), {
      args: ["eval", denoCode],
      stdout: "null",
      stderr: "null",
    }).output();
  }
  if (typeof globalThis.Bun !== "undefined") {
    return globalThis.Bun.spawn(
      [globalThis.process.execPath, "-e", nodeCode],
      { stdout: "ignore", stderr: "ignore" },
    ).exited;
  }
  return new Promise((resolve, reject) => {
    const child = spawn(globalThis.process.execPath, ["-e", nodeCode], {
      stdio: "ignore",
    });
    child.once("error", reject);
    child.once("exit", resolve);
  });
}

export function synchronous_timeout() {
  if (typeof globalThis.Deno !== "undefined") {
    new globalThis.Deno.Command(globalThis.Deno.execPath(), {
      args: [
        "eval",
        `await new Promise(resolve => setTimeout(resolve, 200)); await Deno.writeTextFile(${JSON.stringify(descendantMarker)}, "survived");`,
      ],
      stdout: "null",
      stderr: "null",
    }).outputSync();
    return;
  }
  const code = `setTimeout(() => require("node:fs").writeFileSync(${JSON.stringify(descendantMarker)}, "survived"), 200);`;
  if (typeof globalThis.Bun !== "undefined") {
    globalThis.Bun.spawnSync([globalThis.process.execPath, "-e", code], {
      stdout: "ignore",
      stderr: "ignore",
    });
    return;
  }
  spawnSync(globalThis.process.execPath, ["-e", code], { stdio: "ignore" });
}

export function reset_descendant_marker() {
  if (existsSync(descendantMarker)) unlinkSync(descendantMarker);
}

export function descendant_marker_exists() {
  return existsSync(descendantMarker);
}

export function run_all_crash_cancels_sibling() {
  return true;
}

File: test/runtime_fixture_ffi.erl
-module(runtime_fixture_ffi).
-export([promise_pass/0, promise_reject/0, promise_never/0,
         left_value/0, right_value/0, error_result/0,
         left_string/0, right_string/0, non_binary_assert/0, spawn_descendant/0,
         spawn_cleanup_race/0, spawn_orphan_descendant/0,
         spawn_native_orphan_descendant/0,
         spawn_port_orphan_descendant/0,
         kill_output_collector/0,
         kill_test_owner_from_link/0,
         exit_test_worker/0,
         spawn_native_descendant/0, complete_native_child/0,
         spawn_synchronous_descendant/0,
         native_output_timeout/0, synchronous_timeout/0,
         reset_descendant_marker/0, descendant_marker_exists/0,
         run_all_crash_cancels_sibling/0]).

promise_pass() -> nil.
promise_reject() -> erlang:error(<<"async rejected">>).
promise_never() -> timer:sleep(100), nil.
left_value() -> 1.
right_value() -> 2.
error_result() -> {error, <<"not an integer">>}.
left_string() -> <<"same\nold">>.
right_string() -> <<"same\nnew">>.
non_binary_assert() ->
    erlang:error(#{gleam_error => assert,
                   kind => guard,
                   left => #{value => 1},
                   right => #{value => 2},
                   message => <<"synthetic non-binary assert">>}).

spawn_descendant() ->
    Marker = descendant_marker(),
    spawn(fun() ->
              timer:sleep(100),
              file:write_file(Marker, <<"survived">>)
          end),
    nil.

spawn_cleanup_race() ->
    Marker = descendant_marker(),
    Root = self(),
    spawn_opt(fun() ->
        Reference = erlang:monitor(process, Root),
        receive
            {'DOWN', Reference, process, Root, _Reason} ->
                spawn(fun() ->
                    timer:sleep(50),
                    file:write_file(Marker, <<"survived">>)
                end)
        after 1000 ->
            ok
        end
    end, [{priority, high}]),
    nil.

spawn_orphan_descendant() ->
    Marker = descendant_marker(),
    {_Parent, Reference} = spawn_monitor(fun() ->
        spawn(fun() ->
            timer:sleep(100),
            file:write_file(Marker, <<"survived">>)
        end)
    end),
    receive
        {'DOWN', Reference, process, _Pid, _Reason} -> nil
    after 1000 ->
        erlang:error(orphan_parent_did_not_exit)
    end.

spawn_native_orphan_descendant() -> spawn_orphan_descendant().

spawn_port_orphan_descendant() ->
    Marker = descendant_marker(),
    Erl = os:find_executable("erl"),
    Inner = lists:flatten(io_lib:format(
        "timer:sleep(400), file:write_file(~p, <<\"survived\">>), halt().",
        [Marker])),
    open_port({spawn_executable, Erl},
              [exit_status,
               {args, ["-noshell", "-eval", Inner]}]),
    nil.

kill_output_collector() ->
    exit(group_leader(), kill),
    nil.

kill_test_owner_from_link() ->
    spawn_link(fun() -> exit(kill) end),
    receive after 1000 -> nil end.

exit_test_worker() -> nil.

spawn_native_descendant() -> spawn_descendant().

complete_native_child() ->
    file:write_file(descendant_marker(), <<"completed">>),
    nil.

spawn_synchronous_descendant() -> spawn_descendant().

native_output_timeout() ->
    %% Keep the side effect well beyond the 40 ms isolation timeout. A loaded
    %% scheduler can start this descendant late; the assertion must prove it
    %% was killed, not depend on it writing inside a narrow 210 ms window.
    Marker = descendant_marker(),
    spawn(fun() ->
              timer:sleep(1000),
              file:write_file(Marker, <<"survived">>)
          end),
    timer:sleep(1000),
    nil.

synchronous_timeout() -> native_output_timeout().

reset_descendant_marker() ->
    file:delete(descendant_marker()),
    nil.

descendant_marker_exists() ->
    filelib:is_file(descendant_marker()).

run_all_crash_cancels_sibling() ->
    Caller = self(),
    try
        kangaroo_vm_ffi:run_all([
            fun() -> exit(synthetic_worker_crash) end,
            fun() ->
                timer:sleep(100),
                Caller ! kangaroo_sibling_survived,
                nil
            end
        ])
    catch
        error:{kangaroo_worker_crashed, _Reason} -> ok
    end,
    timer:sleep(150),
    receive
        kangaroo_sibling_survived -> false
    after 0 ->
        true
    end.

descendant_marker() ->
    Temp = temporary_directory(),
    filename:join(Temp, "kangaroo-isolate-descendant-" ++ os:getpid() ++ ".marker").

temporary_directory() ->
    case os:getenv("TMPDIR") of
        false ->
            case os:getenv("TEMP") of
                false ->
                    case os:getenv("TMP") of
                        false -> "/tmp";
                        Value -> Value
                    end;
                Value -> Value
            end;
        Value -> Value
    end.
